docs/design: Upgrade Resource State (#228)
hashicorp/terraform-plugin-framework: docs/design/upgrade-resource-state.md

## Background

The resource state handling between the Terraform CLI and the provider is fairly transparent to practitioners as it is implemented without any particular user interface. Resource state upgrade operations happen as part of the general planning workflow. Practitioners will only have issues if any potential upgrades are incorrectly implemented, such as mismatched types, which will generate errors and likely require further provider action (e.g. a new release with fixed upgrade) or worst case of manual state manipulation.

The next sections will outline some of the underlying details relevant to implementation proposals in this framework.

### Terraform Plugin Protocol

> The specification between Terraform CLI and plugins, such as Terraform Providers, is currently implemented via [Protocol Buffers](https://developers.google.com/protocol-buffers). Below highlights some of the service `rpc` (called by the Terraform CLI) and `message` types that are intergral for upgrade resource state support.

Reviewer: ```suggestion
The specification between Terraform CLI and plugins, such as Terraform Providers, is currently implemented via [Protocol Buffers](https://developers.google.com/protocol-buffers). Highlighted below are some of the service `rpc` (called by the Terraform CLI) and `message` types that are integral for upgrade resource state support.
```

}

// Existing type
type Resource interface {
    // New required method
    // If there is no previous version, a blank implementation would be required.
    UpgradeState(context.Context, /*...*/) /*...*/
}
```

This would ensure framework resources:

- Contain all schema and logic necessary over time to perform any resource state upgrades, which simplifies the framework implementation.
- Do not require additional provider developer discovery for resource versioning features.
- Have access to the provider client during upgrade state operations.

However, there are quite a few drawbacks to this type of implementation:

- Additional and generally unnecessary complexity for most provider developers.
- Lots of code build up over time unless provider developers find ways to reduce it.
- Existing framework provider implementations would require updates to `GetResources` and each `Resource`.
- Confusing blank `(Resource).UpgradeState()` implementation for single version resources. It could be made as an optional method, but then provider developers need to discover it and know when to use it, removing a main benefit of this type of approach.

> ### New ResourceTypeWithUpgrades Interface

Author: ```suggestion
### New ResourceTypeWithUpgradeState Interface
```

The framework could allow provider developers to optionally extend their `ResourceType` with a new `UpgradeState` method.

For example in the framework:

```go
// New type
type ResourceTypeWithUpgradeState interface {
    UpgradeState(context.Context, /*...*/) /*...*/
}
```

Which would result in the following optional provider implementation:

```go
func (rt ExampleResourceType) UpgradeState(ctx context.Context, /*...*/) /*...*/ {
    /* ... */
}
```

> However, there is an immediate drawback that there would be no access to the provider client during upgrade state operations. This is considered a non-starter due to the goals of this functionality.

Reviewer: Although the headings are different, this text is identical to the following section - _New ResourceWithUpgradeState Interface_.

Author: Hey sorry, I'm not sure what you mean here. 😅  `ResourceType` and `Resource` are very different implementation details. Providers get the opportunity to inject their clients into `Resource`, but not `ResourceType`.

Reviewer: My bad 😄

```go
// New type
type ResourceWithUpgradeState interface {
    UpgradeState(context.Context, UpgradeResourceStateRequest, *UpgradeResourceStateResponse)
}

// New type
type UpgradeResourceStateRequest struct {
    // Current state version
    // May be worth renaming to CurrentStateVersion in the actual implementation
    Version int64

    // JSON encoded or flatmap state
    // The type would likely be abstracted in the framework, but showing for brevity.
    RawState tfprotov6.RawState
}

// New type
type UpgradeResourceStateResponse struct {
    // Upgraded state, which must be sent as msgpack
    // The underlying *tftypes.DynamicValue type is very unfriendly for
    // provider developers at this layer so a next option is to rely on
    // the existing State type.
    State State

    Diagnostics diag.Diagnostics
}
```

Which would result in the following optional provider implementation:

```go
func (r ExampleResource) UpgradeState(context.Context, req tfsdk.UpgradeResourceStateRequest, *tfsdk.UpgradeResourceStateResponse) {
    // potentially branching logic based on req.Version
    // complex logic to convert req.RawState into the resp.State
}
```

> In this setup, the framework could only implement a thin abstraction over the request data since it has no previous schema information. One framework workaround to remove this hurdle for provider developers might be to introduce additional method on `ResourceType` or `Resource` to fetch previous schema.

Reviewer 2: introduce additional method -> introduce an additional method

Author: ```suggestion
In this setup, the framework could only implement a thin abstraction over the request data since it has no previous schema information. One framework workaround to remove this hurdle for provider developers might be to introduce an additional method on `ResourceType` or `Resource` to fetch previous schema.
```

    RawState tfprotov6.RawState

    // Populated if ResourceStateUpgrader.Schema is present
    State State
}

// New type
type UpgradeResourceStateResponse struct {
    State State

    Diagnostics diag.Diagnostics
}
```

Which would result in the following optional provider implementation:

```go
func (r ExampleResource) UpgradeState(ctx context.Context) map[int64]tfsdk.ResourceStateUpgrader {
    return map[int64]tfsdk.ResourceStateUpgrader{
        0: {
            StateUpgrader: func(ctx context.Context, req tfsdk.UpgradeResourceStateRequest, resp tfsdk.UpgradeResourceStateResponse) {
                // logic to handle req.RawState
                // resp.State = ...
            },
        },
        1: {
            Schema: tfsdk.Schema{ /* ... */ },
            StateUpgrader: func(ctx context.Context, req tfsdk.UpgradeResourceStateRequest, resp tfsdk.UpgradeResourceStateResponse) {
                // logic to handle req.State
                // resp.State = ...
            },
        }
    }
}
```

Whether to implement separate methods for state upgrade logic is up to the provider developer.

This would ensure framework resources:

> - Succintly define state upgrade information in one place (it is a provider developer decision whether to add non-framework methods or other coding techniques to fully expand the details).

Reviewer: ```suggestion
- Succinctly define state upgrade information in one place (it is a provider developer decision whether to add non-framework methods or other coding techniques to fully expand the details).
```
